packages/test/test-end-to-end-tests/src/test/SummarizerWithDataStoreRealization.spec.ts

```diff
@@ -0,0 +1,329 @@
+  ? never // exclude indexers [key:string |number]: any
+  : TProp extends keyof T[TProp] // TProp is a property of T, T[TProp] and, T[TProp][TProp]
+  ? TProp extends keyof T[TProp][TProp] // ex; IProvideFoo.IFoo.IFoo.IFoo
+    ? TProp
+    : never
+  : never;
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+type Provider<T extends Object = Object> = Partial<Pick<T, ProviderPropertyKeys<T>>>;
+
+interface ProvideSearchContent {
+    SearchContent: SearchContent;
+  }
+interface SearchContent extends ProvideSearchContent {
+    getSearchContent(): Promise<string | undefined>;
+  }
+
+// Note GC needs to be disabled.
+const runtimeOptions: IContainerRuntimeOptions = {
+    summaryOptions: {
+        disableSummaries: true,
+        summaryConfigOverrides: { state: "disabled" },
+    },
+    gcOptions: { gcAllowed: false },
+};
+
+export const TestDataObjectType1 = "@fluid-example/test-dataStore1";
+export const TestDataObjectType2 = "@fluid-example/test-dataStore2";
+class TestDataObject2 extends DataObject implements SearchContent {
+    public async getSearchContent(): Promise<string | undefined> {
+        return Promise.resolve("TestDataObject2 Search Blob");
+    }
+    public get SearchContent() {
+        return this;
+    }
+    public get _root() {
+        return this.root;
+    }
+    public get _context() {
+        return this.context;
+    }
+    private readonly mapKey = "SharedMap";
+    public map!: SharedMap;
+
+    protected async initializingFirstTime() {
+        const sharedMap = SharedMap.create(this.runtime, this.mapKey);
+        this.root.set(this.mapKey, sharedMap.handle);
+   }
+
+    protected async hasInitialized() {
+        const mapHandle = this.root.get<IFluidHandle<SharedMap>>(this.mapKey);
+        assert(mapHandle !== undefined, "SharedMap not found");
+        this.map = await mapHandle.get();
+    }
+}
+
+class TestDataObject1 extends DataObject implements SearchContent {
+    public async getSearchContent(): Promise<string | undefined> {
+        // By this time, we are in the middle of the summarization process and
+        // the DataStore should have been initialized with no child.
+        // We will force it to be realized so when we invoke completeSummary on the SummarizerNode it would
+        // cause bug https://dev.azure.com/fluidframework/internal/_workitems/edit/1633 to happen.
+        const dataTestDataObject2Handle = this.root.get<IFluidHandle<TestDataObject2>>("dsFactory2");
+        assert(dataTestDataObject2Handle, "dsFactory2 not located");
+        const dataStore2 = await dataTestDataObject2Handle.get();
+        dataStore2.map.set("mapkey", "value");
+
+        return Promise.resolve("TestDataObject1 Search Blob");
+    }
+    public get SearchContent() {
+        return this;
+    }
+    public get _root() {
+        return this.root;
+    }
+
+    public get _context() {
+        return this.context;
+    }
+
+    private readonly matrixKey = "SharedMatrix";
```

Comment:
Matrix, counter, undo redo stack manager are not really needed. You can remove these to make the data object simple. In the test, you can update the root map instead of using the matrix.

Reply:
Basically, TestDataObject1 is similar to TestDataObject2 expect that it creates and loads TestDataObject2.

Reply:
I Like adding the elements as it makes it easier to debug and understand which DataStore is which.